pkg/identity/identitysync/one-login.go

```diff
@@ -0,0 +1,333 @@
+package identitysync
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/common-fate/granted-approvals/pkg/gconfig"
+	"github.com/common-fate/granted-approvals/pkg/identity"
+	"github.com/pkg/errors"
+)
+
+const OneLoginBaseURL = "https://commonfate-dev.onelogin.com"
+
+type OneLoginSync struct {
+	// This is initialised during the Init function call and is not saved in config
+	clientID     gconfig.StringValue
+	clientSecret gconfig.SecretStringValue
+	token        gconfig.SecretStringValue
+}
+
+func (s *OneLoginSync) Config() gconfig.Config {
+	return gconfig.Config{
+		gconfig.StringField("clientId", &s.clientID, "the One Login client ID"),
+		gconfig.SecretStringField("clientSecret", &s.clientSecret, "the One Login client secret", gconfig.WithNoArgs("/granted/secrets/identity/one-login/secret")),
+	}
+}
+
+func (s *OneLoginSync) Init(ctx context.Context) error {
+
+	url := OneLoginBaseURL + "/auth/oauth2/v2/token"
+
+	var jsonStr = []byte(`{ "grant_type": "client_credentials"}`)
+	req, _ := http.NewRequest("POST", url, bytes.NewBuffer(jsonStr))
+	req.Header.Add("Authorization", "client_id: "+s.clientID.Get()+",client_secret: "+s.clientSecret.Get())
+	req.Header.Add("content-type", "application/json")
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+	b, err := io.ReadAll(res.Body)
+	if err != nil {
+		return err
+	}
+
+	//return the error if its anything but a 200
+	if res.StatusCode != 200 {
+		return fmt.Errorf(string(b))
+	}
+
+	var lu GetAccessTokenResponse
+	err = json.Unmarshal(b, &lu)
+	if err != nil {
+		return err
+	}
+
+	s.token.Set(lu.AccessToken)
+
+	return nil
+}
+func (s *OneLoginSync) TestConfig(ctx context.Context) error {
+	_, err := s.ListUsers(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to list users while testing one login identity provider configuration")
+	}
+	_, err = s.ListGroups(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to list groups while testing one login identity provider configuration")
+	}
+	return nil
+}
+
+// userFromOktaUser converts a Okta user to the identityprovider interface user type
+func (o *OneLoginSync) idpUserFromOneLoginUser(ctx context.Context, oneLoginUser *OneLoginUser) (identity.IDPUser, error) {
+	u := identity.IDPUser{
+		ID:        strconv.Itoa(oneLoginUser.ID),
+		FirstName: oneLoginUser.Firstname,
+		LastName:  oneLoginUser.Lastname,
+		Email:     oneLoginUser.Email,
+		Groups:    []string{},
+	}
+
+	for _, r := range oneLoginUser.RoleID {
+		u.Groups = append(u.Groups, strconv.Itoa(r))
+	}
+
+	return u, nil
+}
+
+func (o *OneLoginSync) idpGroupFromOneLoginGroup(oneLoginGroup OneLoginGroup) identity.IDPGroup {
+	return identity.IDPGroup{
+		ID:   strconv.Itoa(oneLoginGroup.ID),
+		Name: oneLoginGroup.Name,
+	}
+
+}
+
+func (a *OneLoginSync) ListUsers(ctx context.Context) ([]identity.IDPUser, error) {
+
+	//get all users
+	idpUsers := []identity.IDPUser{}
```

Comment:
nitpick: initialise a zero value for a variable
```suggestion
	var idpUsers []identity.IDPUser
```